src/components/forms/DateInput.jsx

```diff
@@ -0,0 +1,212 @@
+import React, { useState } from 'react';
+import PropTypes from 'prop-types';
+import moment from 'moment';
+
+import DayPicker from 'react-day-picker';
+
+import ActionsArrowLeftIcon from 'lib/icons/react/ActionsArrowLeftIcon';
+import ActionsArrowRightIcon from 'lib/icons/react/ActionsArrowRightIcon';
+import Popover from '../popovers/Popover';
+import Flex from '../layout/Flex';
+import FlexItem from '../layout/FlexItem';
+import IconButton from '../interactive/IconButton';
+
+const DATE_FORMAT = 'MM/DD/YYYY';
+const DATE_PATTERN = '[0-9/]*';
+
+// tests for full date string in either MM/DD/YYYY or M/D/YYYY format
+const VALID_DATE_REGEX = /^(0?[1-9]|1[0-2])[/](0?[1-9]|[12]\d|3[01])[/]\d{4}$/;
+
+/**
+ * Returns a range of years to show in date picker. The range will expand to include
+ * the selected date, if a date has been selected.
+ *
+ * @param {Number} currentYear YYYY
+ * @param {Number} pastYears number of prior years in range
+ * @param {Number} futureYears number of future years in range
+ * @param {Date} selectedDate currently selected date from DateInput
+ * @returns {Object} { startYear: YYYY, endYear: YYYY }
+ */
+export const getYearRange = (currentYear, pastYears, futureYears, selectedDate) => {
+  let startYear = new Date(currentYear - pastYears, 0).getFullYear();
+  let endYear = new Date(currentYear + futureYears + 1, 11).getFullYear();
+  const selectedYear = selectedDate instanceof Date && selectedDate.getFullYear();
+
+  // Expand range to include selected year if out of range
+  if (selectedYear && selectedYear < startYear) {
+    startYear = selectedYear;
+  }
+  if (selectedYear && selectedYear > endYear) {
+    endYear = selectedYear + 1;
+  }
+
+  return { startYear, endYear };
+};
```

Comment:
This is personal preference maybe, but functions like this that have this many arguments, it's usually not immediately clear what the function is doing in use, vs if you use an object.

e.g.

```
getYearRange(2020, 3, 5) // I have no idea what 3 and 5 are, without knowing that the function signature is
getYearRange({ year: 2020, pastYears: 3, futureYears: 5}) // this is more clear on glance
getYearRange(2020, { pastYears: 3, futureYears: 5}) // this is also somewhat clear as well.
``

Reply:
Yeah, I agree with this generally - 4 positional arguments is pretty much the mental limit when I break out args into object properties. If this were exposed in docs and intended for usage outside of this component, I would definitely favor an object here.

In this case, this function covers logic that I factored out of `DateInput` to make it a pure function. Above all else, I wanted this function to be "pure". I preferred positional arguments to do this because...

- only the last param is optional
- explicitly declaring the arguments keeps the surface area of the function interface as small as possible (objects can contain _anything_)
- accepting an object means you could have unrelated parameters being passed in
- positional args discourage doing "unpure" things like mutating the object being passed in
- easier to jsDoc params this way

Reply:
*tl;dr* - positional arguments are "safer" if functional purity is the goal, object arguments are nicer for public interfaces. `getYearRange` is not intended to be a public interface